### PR TITLE
[ddp] Support multi-node distributed execution under torchelastic

### PR DESCRIPTION
**Pull request by ashwinb**
## What does this PR do?

It allows using Lightning distributed trainers where workers are managed by torchelastic (https://github.com/pytorch/elastic)

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -943,11 +943,13 @@ def init_ddp_connection(
             os.environ['MASTER_PORT'] = '12910'
         log.debug(f"MASTER_PORT: {os.environ['MASTER_PORT']}")
 
-        if 'WORLD_SIZE' in os.environ and os.environ['WORLD_SIZE'] != world_size:
-            log.warning("WORLD_SIZE environment variable is not equal to the computed "
-                        "world size. Ignored.")
+        if 'WORLD_SIZE' in os.environ and int(os.environ['WORLD_SIZE']) != world_size:
+            log.warning("WORLD_SIZE environment variable ({}) is not equal to the computed "
```
Comment (tullie) [inline pytorch_lightning/core/lightning.py:947]: Can you please use python format string literals instead f""

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 2]
```diff
@@ -305,13 +305,19 @@ def ddp_train(self, process_idx, model):
         :param cluster_obj:
         :return:
         """
-        # node rank using relative slurm id if under slurm management
+        # if under slurm management, use SLURM_NODEID,
+        # if torchelastic, use GROUP_RANK,
         # otherwise use given node rank or default to node rank 0
-        try:
-            node_id = os.environ['SLURM_NODEID'] if self.is_slurm_managing_tasks else os.environ['NODE_RANK']
+        node_keys = ['SLURM_NODEID', 'NODE_RANK', 'GROUP_RANK']
```
Comment (tullie) [inline pytorch_lightning/trainer/distrib_data_parallel.py:311]: I'm concerned about cases when multiple of these are set. Can we log a warning if we use one of these IDs and the other two are set? I think writing the code that way would make it easier to understand what's happening here too.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -796,11 +796,14 @@ def fit(
         if self.use_ddp2:
             task = int(os.environ['SLURM_LOCALID'])
             self.ddp_train(task, model)
-
         elif self.use_ddp:
             if self.is_slurm_managing_tasks:
                 task = int(os.environ['SLURM_LOCALID'])
                 self.ddp_train(task, model)
+            # torchelastic
+            elif 'WORLD_SIZE' in os.environ and 'GROUP_RANK' in os.environ:
```
Comment (tullie) [inline pytorch_lightning/trainer/trainer.py:804]: I wonder what we need to do to get this working on slurm. It could be as simple as using the LOCAL_RANK environment instead of the SLURM_LOCALID.

I'll look into it.

Comment (tullie) [reply pytorch_lightning/trainer/trainer.py:804]: Okay so looked into it and it's not that simple. Basically, much like distributed training, there's a few ways to initialize elastic training. However, because elastic training needs to own the processes to work, slurm can't spawn them for it.

In distributed training you have the options:
1. run ```python -m torch.distributed.launch <train_script.py>```, which creates the processes for you
2. start the processes yourself using mp (e.g. in the else of this if statement)
3. let slurm (or another scheduler) create the processes

In elastic training you have the options:
1. run ```python3 -m torchelastic.distributed.launch```,  which creates the processes and handles the fault-tolerence and elastic workers.
2. create an elastic agent such as [LocalElasticAgent](https://pytorch.org/elastic/0.2.0/agent.html#torchelastic.agent.server.local_elastic_agent.LocalElasticAgent). This will spawn the elastic processes and manage them with the synchronous function LocalElasticAgent.run().

This doesn't leave a particularly easy way to do slurm because the agent needs to spawn the processes. My guess is that you need to configure slurm to have 1 process per node (i.e. ntasks-per-node=1) and then create the agent and processes as explained in 2. at the beginning of training. You'd also need to setup the distributed key-value store backend (Etcd or Zeus). Luckily they've provided a helpful [python API for spawning Etcd server](https://pytorch.org/elastic/0.2.0/rendezvous.html#torchelastic.rendezvous.etcd_server.EtcdServer).

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -943,11 +943,13 @@ def init_ddp_connection(
...
+                        "world size ({}). Ignored.".format(
+                            os.environ['WORLD_SIZE'], world_size))
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:949]:
```suggestion
            log.warning(f"WORLD_SIZE environment variable ({os.environ['WORLD_SIZE']})"
                        f"is not equal to the computed world size ({world_size}). Ignored.")
```

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 5]
```diff
@@ -305,13 +305,19 @@ def ddp_train(self, process_idx, model):
...
+            node_id = os.environ.get(k, None)
+            if node_id:
+                break
+        if node_id is not None:
             self.node_rank = int(node_id)
-        except KeyError:
-            log.warning("SLURM_NODEID or NODE_RANK environment variable is not defined. Set as 0.")
+        else:
+            log.warning("SLURM_NODEID, NODE_RANK or GROUP_RANK environment "
+                        "variables not defined. Set as 0.")
```
Comment (Borda) [inline pytorch_lightning/trainer/distrib_data_parallel.py:320]:
```suggestion
            log.warning("SLURM_NODEID, NODE_RANK or GROUP_RANK environment"
                        " variables not defined. Set as 0.")
```

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/distrib_data_parallel.py:311]: good idea.

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (ashwinb) [reply pytorch_lightning/trainer/trainer.py:804]: @tullie I don't know what you mean. Lightning already works correctly under a Slurm managed task environment. Do you mean having the same code for both pytorch elastic and slurm?

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py:804]: lightning already spawns workers when it detects it is not being managed by slurm

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L801-L813

@tullie is this what you mean?

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py:804]: image

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py:804]: yeah, this should work fine no?
If this is the route we take with elastic it means that something else created the process that called each script. Is that the expected behavior @ashwinb @tullie (I haven't used elastic yet).

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -483,8 +483,8 @@ def __init__(
         # init flags for SLURM+ddp to work
         self.proc_rank = 0
         self.world_size = 1
-        self.node_rank = 0
         self.configure_slurm_ddp(self.num_nodes)
+        self.node_rank = self.determine_node_rank()
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/trainer.py:487]: seems like this function is not defined?

Comment (ashwinb) [reply pytorch_lightning/trainer/trainer.py:487]: @williamFalcon it is! in the `distributed_data_parallel.py` mixin.

Comment (ashwinb) [reply pytorch_lightning/trainer/trainer.py:487]: Ah, typo. Correcting.

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py:487]: oh cool.. ic.

I guess the names don't match
determine_ddp_node_rank -> determine_node_rank

tests are failing there

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py:487]:
```suggestion
        self.node_rank = self. determine_ddp_node_rank()
```

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (ashwinb) [reply pytorch_lightning/trainer/trainer.py:804]: Yes, elastic launches agents on each node which manage the individual worker processes. Lightning's job in that case is to init its process group and configure and run a single trainer worker. This is just like Slurm.

**Comment by williamFalcon** [comment]
@Borda last PR then let's do rc2

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (tullie) [reply pytorch_lightning/trainer/trainer.py:804]: All I was saying is that this PR doesn't add support for Elastic Pytorch in a Slurm managed environment. This is fine for now but ideally they'd be able to work together in the future.

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 7]
```diff
@@ -279,6 +279,24 @@ def configure_slurm_ddp(self, num_gpu_nodes):
...
+        # torchelastic uses the envvar GROUP_RANK, whereas other systems(?) use NODE_RANK.
+        # otherwise use given node rank or default to node rank 0
+        env_vars = ['NODE_RANK', 'GROUP_RANK']
+        node_ids = [(k, os.environ.get(k, None)) for k in env_vars]
+        node_ids = [(k, v) for k, v in node_ids if v is not None]
+        if len(node_ids) == 0:
+            log.warning("No environment variable for node rank defined. Set as 0.")
+            return 0
+        if len(node_ids) > 1:
+            log.warning(f"Multiple environment variables ({keys(node_ids)} defined for node rank. "
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/distrib_data_parallel.py:295]: "keys" are undefined here

Comment (williamFalcon) [reply pytorch_lightning/trainer/distrib_data_parallel.py:295]:
```
./pytorch_lightning/trainer/distrib_data_parallel.py:295: [F821] undefined name 'keys'
```
